src/plots/cartesian/axes.js

```diff
@@ -571,34 +571,44 @@ axes.calcTicks = function calcTicks(ax) {
     if((ax._tmin < startTick) !== axrev) return [];
 
     // return the full set of tick vals
-    var vals = [];
+    var tickVals = [];
     if(ax.type === 'category' || ax.type === 'multicategory') {
         endTick = (axrev) ? Math.max(-0.5, endTick) :
             Math.min(ax._categories.length - 0.5, endTick);
     }
 
+    var isLog = (ax.type === 'log');
+
     var xPrevious = null;
     var maxTicks = Math.max(1000, ax._length || 0);
     for(var x = ax._tmin;
             (axrev) ? (x >= endTick) : (x <= endTick);
             x = axes.tickIncrement(x, ax.dtick, axrev, ax.calendar)) {
         // prevent infinite loops - no more than one tick per pixel,
         // and make sure each value is different from the previous
-        if(vals.length > maxTicks || x === xPrevious) break;
+        if(tickVals.length > maxTicks || x === xPrevious) break;
         xPrevious = x;
 
-        vals.push(x);
+        var minor = false;
+        if(isLog && (x !== (x | 0))) {
```

Comment:
What does `x !== (x | 0)` mean?

Reply:
Ha I see, it means _if `x` isn't an integer_, right?

Reply:
In https://github.com/plotly/plotly.js/issues/3250 , @alexcjohnson wrote

> seems like for dtick='D1' and 'D2', tickprefix (and ticksuffix if given) should only apply to the "complete" labels at the powers of 10, not the single-digit labels between.

@archmoj can you confirm that _should only to "complete" labels at the powers of 10" and _if `x` isn't an integer_ are equivalent here?

Reply:
> Ha I see, it means _if `x` isn't an integer_, right?

Correct.

Reply:
> In #3250 , @alexcjohnson wrote
> 
> > seems like for dtick='D1' and 'D2', tickprefix (and ticksuffix if given) should only apply to the "complete" labels at the powers of 10, not the single-digit labels between.
> 
> @archmoj can you confirm that _should only to "complete" labels at the powers of 10" and _if `x` isn't an integer_ are equivalent here?

Yes. It seems so.

Reply:
Have you tried your branch together with `dtick` of the `L<n>` and `D<n>` families?

https://github.com/plotly/plotly.js/blob/56dd829fbaa2ba5e254bc6870b380ef333bf30e5/src/plots/cartesian/layout_attributes.js#L296-L300

Reply:
> Have you tried your branch together with `dtick` of the `L<n>` and `D<n>` families?
> 
> https://github.com/plotly/plotly.js/blob/56dd829fbaa2ba5e254bc6870b380ef333bf30e5/src/plots/cartesian/layout_attributes.js#L296-L300

@etpinard Good call! 
Fixed new logic & improved the test in https://github.com/plotly/plotly.js/pull/4064/commits/27067a6c8f45dc0fbfa19dbaad0393047773601c.